Utils.ps1

```diff
@@ -104,6 +140,281 @@ function Set-LabAccessControl {
+      }
+  
+      $vnetAddressRangeStart = $vnetAddressRangeInfo.NetworkAddress
+      $nextVnetAddressRangeStart = (($vnetAddressRangeInfo.Broadcast | Convert-IPToDecimal) + 1) | Convert-DecimalToIp 
+  
+      $allocatedSubnetRanges.Add($vnetAddressRangeStart, $vnetAddressRangeStart)
+      $allocatedSubnetRanges.Add($nextVnetAddressRangeStart, $nextVnetAddressRangeStart)
+  
+      $vnetSubnets = $VirtualNetwork.Subnets  | `
+                      Sort-Object -Property AddressPrefix
+      $vnetSubnets | Foreach-Object {
+  
+        $vnetSubnet = $_
+        $subnetAddressRangeInfo = Invoke-PSipcalc -NetworkAddress $vnetSubnet.AddressPrefix
+        $subnetAddressRangeLength = $subnetAddressRangeInfo.NetworkLength
+        
+        if ($subnetAddressRangeLength -gt $Length) {
+          # This subnet is not large enough to host a new subnet
+          continue
+        }
+  
+        $subnetAddressRangeStart = $subnetAddressRangeInfo.NetworkAddress
+        $nextsubnetAddressRangeStart = (($subnetAddressRangeInfo.Broadcast | Convert-IPToDecimal) + 1) | Convert-DecimalToIp
+  
+        if ($allocatedSubnetRanges[$subnetAddressRangeStart]) {
+          # Replace existing with the new range
+          $allocatedSubnetRanges[$subnetAddressRangeStart] = $nextsubnetAddressRangeStart
+        }
+        else {
+          $allocatedSubnetRanges.Add($subnetAddressRangeStart, $nextsubnetAddressRangeStart)
+        }
+      }
+    }
+
+    # Ordered Array of unallocated ranges (CIDR notation) calculated as a complement to the above table
+    $unallocatedSubnetRanges = [System.Collections.ArrayList]@()
+
+    $allocatedSubnetStarts = [array]$allocatedSubnetRanges.Keys
+    $allocatedSubnetStarts | ForEach-Object {
+
+      $index = $allocatedSubnetStarts.IndexOf($_)
+      $nextsubnetAddressRangeStart = $allocatedSubnetRanges[$_]
+      if ($index+1 -lt $allocatedSubnetRanges.Count) {
+
+        $nextNextsubnetAddressRangeStart = $allocatedSubnetStarts[$index+1]
+        if ($nextsubnetAddressRangeStart -ne $nextNextsubnetAddressRangeStart) {
+          
+          # Calculate the end IP of the allocated range
+          $nextsubnetAddressRangeEnd = (($nextNextsubnetAddressRangeStart | Convert-IPToDecimal) - 1) | Convert-DecimalToIp 
+          
+          $unallocatedAddressRange = Convert-IPRangeToCIDRNotation -Start $nextsubnetAddressRangeStart -End $nextsubnetAddressRangeEnd 
+          $unallocatedSubnetRanges.Add($unallocatedAddressRange) | Out-Null
+
+        } # else contiguous subnet
+      }
+    }
+    
+    # TODO deciding on the smallest (bottom) or shortest subnet
+    $unallocatedSubnetRanges | Select-Object -First 1
+  }
+  catch {
+    Write-Error -ErrorRecord $_
+  }
+  finally {
+    Remove-PsipcalcScript
+  }
+}
+
+# Get the first available subnet with a unassigned address space large enough to host a new subnet of at least /$Length
+function Get-VirtualNetworkUnassignedSpace {
```

Comment:
Get the first available subnet with a unassigned address space large enough to host a new subnet of /$Length